pandas/_libs/tslibs/np_datetime.pyx

```diff
@@ -273,14 +274,25 @@ cdef inline int string_to_dts(
     int* out_local,
     int* out_tzoffset,
     bint want_exc,
+    format: str | None = None,
+    bint exact = True,
 ) except? -1:
     cdef:
         Py_ssize_t length
         const char* buf
+        Py_ssize_t format_length
+        const char* format_buf
 
     buf = get_c_string_buf_and_size(val, &length)
```

Comment:
Orthogonal to this but if you want to work more with the C codebase I think we should just replace `get_c_string_buf_and_size` with `PyUnicode_AsUTF8AndSize` directly; the former might have served a purpose with the Py2/3 transition but is just an unnecessary layer at this point